Modules/pyexpat.c

```diff
@@ -1631,101 +1627,152 @@ static int init_handler_descrs(void)
     return 0;
 }
 
-PyMODINIT_FUNC
-MODULE_INITFUNC(void)
+static int
+pyexpat_traverse(PyObject *module, visitproc visit, void *arg)
+{
+    pyexpat_state *state = pyexpat_get_state(module);
+    Py_VISIT(state->xml_parse_type);
+    Py_VISIT(state->error);
+    return 0;
+}
+
+static int
+pyexpat_clear(PyObject *module)
+{
+    pyexpat_state *state = pyexpat_get_state(module);
+    Py_CLEAR(state->xml_parse_type);
+    Py_CLEAR(state->error);
+    return 0;
+}
+
+static void
+pyexpat_free(void *module)
+{
+    pyexpat_clear((PyObject *)module);
+}
+
+static int
+pyexpat_exec(PyObject *mod)
 {
-    PyObject *m, *d;
     PyObject *errmod_name = PyUnicode_FromString(MODULE_NAME ".errors");
-    PyObject *errors_module;
-    PyObject *modelmod_name;
-    PyObject *model_module;
-    PyObject *tmpnum, *tmpstr;
-    PyObject *codes_dict;
-    PyObject *rev_codes_dict;
-    int res;
-    static struct PyExpat_CAPI capi;
-    PyObject *capi_object;
+    if (errmod_name == NULL) {
+        return -1;
+    }
 
-    if (errmod_name == NULL)
-        return NULL;
-    modelmod_name = PyUnicode_FromString(MODULE_NAME ".model");
-    if (modelmod_name == NULL)
-        return NULL;
+    PyObject *modelmod_name = PyUnicode_FromString(MODULE_NAME ".model");
+    if (modelmod_name == NULL) {
+        return -1;
```

Comment:
return starting near this line leak errmod_name reference. You should add an "error:" label to clear errmod_name and modelmod_name. Something like (fix the indent):

```
errror:
Py_DECREF(errmod_name);
Py_DECREF(modelmod_name);
return -1;
```

and then replace `return -1` with `goto error`.

You need to replace  Py_DECREF(errmod_name); above with  Py_CLEAR(errmod_name).

Or maybe declare these variables above, closer to where they are used.